### PR TITLE
JDK-8273038: ChannelInputStream.transferTo() uses FileChannel.transferTo(FileChannel)

### PR DESCRIPTION
**Pull request by mkarg**
As proposed in JDK-8265891, this PR provides an implementation for `Channels.newInputStream().transferTo()` which provide superior performance compared to the current implementation. This PR is a spin-off from https://github.com/openjdk/jdk/pull/4263 and https://github.com/openjdk/jdk/pull/5179 and deliberately concentrates **only** on the case where both, source **and** target, are actually `FileChannel`s. Other types of channels will be discussed in future PRs.

* Prevents transfers through the JVM heap as much as possibly by offloading to deeper levels via NIO, hence allowing the operating system to optimize the transfer.

Using JMH I have benchmarked both, the original implementation and this implementation, and (depending on the used hardware and use case) performance change was approx. doubled performance. A rather similar approach in different use casse was recently implemented by https://github.com/openjdk/jdk/pull/5097 which was reported by even provide 5x performance (https://github.com/openjdk/jdk/pull/5097#issuecomment-897271997).

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8273038](https://bugs.openjdk.java.net/browse/JDK-8273038): ChannelInputStream.transferTo() uses FileChannel.transferTo(FileChannel)

### Reviewers
 * [Alan Bateman](https://openjdk.java.net/census#alanb) (@AlanBateman - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/5209/head:pull/5209` \
`$ git checkout pull/5209`

Update a local copy of the PR: \
`$ git checkout pull/5209` \
`$ git pull https://git.openjdk.java.net/jdk pull/5209/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 5209`

View PR using the GUI difftool: \
`$ git pr show -t 5209`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/5209.diff">https://git.openjdk.java.net/jdk/pull/5209.diff</a>

</details>

### CODE REVIEW

**Comment by mkarg** [comment]
@AlanBateman @bplb This PR is a spin-off for the single case of FileChannel->FileChannel only, as suggested by you. I kindly request your review. Thanks.

File: test/jdk/sun/nio/ch/ChannelInputStream/TransferTo.java [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+    public interface Thrower {
+        public void run() throws Throwable;
+    }
+
+    public static void assertThrowsNPE(Thrower thrower, String message) {
+        assertThrows(thrower, NullPointerException.class, message);
+    }
+
+    public static <T extends Throwable> void assertThrows(Thrower thrower, Class<T> throwable, String message) {
+        Throwable thrown;
```
Comment (pirateskipper) [inline test/jdk/sun/nio/ch/ChannelInputStream/TransferTo.java:181]: ` Throwable thrown = null;
        try {
            thrower.run();`

**Comment by AlanBateman** [comment]
There are 3 PRs open, I'll ignore #4263 and #5179 for now.

The moving of the OutputStream implementation from Channels to sun.nio.ch.ChannelOutputStream looks fine. In the class description it should be "accessible" as it visible. You can drop "but not be part of the java.base module API" too.

Changing the bounds check to use Objects.checkFromIndexSize looks fine.

It would be good if you could re-format 1-arg transferTo to something like this:
```
        if (out instanceof ChannelOutputStream cos
                && ch instanceof FileChannel fc 
                && cos.channel() instanceof FileChannel dst) {
            
        }
```
just to to avoid the really long line as that gets annoying when looking at side-by-side diffs.

The 2-arg transferTo methods look okay. An alternative, that avoids the need for "srcPos + bytesWritten" in 3 places is to just something like:
```
        long initialPos = src.position();
        long pos = initialPos;
        try {
            while (pos < src.size()) {
               pos += src.transferTo(pos, Long.MAX_VALUE, dst);
            }
        } finally {
            src.position(pos);
        }
        return pos - initialPos;
```

@bplb You looked at the test in the first PR, have you look at this one? I suspect it would be a lot cleaner/maintainable if it were written as a TestNG test. I assume the transfer loop never has more than one iteration because the file sizes are <2GB. If it were writing this test then I would probably use Files/Arrays.mismatch to compare the contents.

**Comment by mkarg** [comment]
> There are 3 PRs open, I'll ignore #4263 and #5179 for now.

Yes please ignore them for now. The other two PRs are suspendend as they cover bigger pictures and will be resumed by me *later*.

> The moving of the OutputStream implementation from Channels to sun.nio.ch.ChannelOutputStream looks fine. In the class description it should be "accessible" as it visible. You can drop "but not be part of the java.base module API" too.

Done in https://github.com/openjdk/jdk/pull/5209/commits/f35cd7d74f2ac15f60e7d97395ec71deefeafb76.

> It would be good if you could re-format 1-arg transferTo to something like this:
> ...
> just to to avoid the really long line as that gets annoying when looking at side-by-side diffs.

Never had this problem as I'm using inline-diff tools only, but done in https://github.com/openjdk/jdk/pull/5209/commits/33bef3d6ef33bb9eddc66e69c1a1e6fd9d565f34.

> An alternative, that avoids the need for "srcPos + bytesWritten" in 3 places is to just something like:
> ...

Does that mean that I must change it? As the method is mostly I/O bound I do not see what we actually would gain.

> I suspect it would be a lot cleaner/maintainable if it were written as a TestNG test. I assume the transfer loop never has more than one iteration because the file sizes are <2GB. If it were writing this test then I would probably use Files/Arrays.mismatch to compare the contents.

Understood, but please note that this code actually was simply copied from the original tests that Brian provided as a blue-print, which neither uses TestNG nor Files/Arrays.mismatch nor tests larger files. So if you ask me to change that, in turn someone should change the original tests then to keep the code in sync. So can we keep the tests as-is or am I forced to use TestNG and File/Arrays.mismatch?

**Comment by bplb** [comment]
@AlanBateman It does not look like this test is substantively different than the one in the first PR aside from constraining the cases tested. One thing I think is insufficient is only dealing with streams at position zero. In the test I recently added, [java/io/FileInputStream/TransferTo.java](https://github.com/openjdk/jdk/blob/master/test/jdk/java/io/FileInputStream/TransferTo.java), I included random initial positions for both input and output. For random positions it is better to use `jdk.test.lib.RandomFactory.getRandom()` instead of `new Random()` as this will print the seed value which can be useful in reproducing any problems. (For data content alone, `new Random()` is fine because we don't care what those bytes are.)

**Comment by AlanBateman** [comment]
> > There are 3 PRs open, I'll ignore #4263 and #5179 for now.
> 
> Yes please ignore them for now. The other two PRs are suspendend as they cover bigger pictures and will be resumed by me _later_.

Okay. The implementation changes in the current patch (205b9b21) look good. I think we will need to iterate on the test.

**Comment by mkarg** [comment]
> > > There are 3 PRs open, I'll ignore #4263 and #5179 for now.
> > 
> > 
> > Yes please ignore them for now. The other two PRs are suspendend as they cover bigger pictures and will be resumed by me _later_.
> 
> Okay. The implementation changes in the current patch ([205b9b2](https://github.com/openjdk/jdk/commit/205b9b21e003937f91311eb91d7671e0810684bb)) look good. I think we will need to iterate on the test.

Ok so I will do as Brian said and include a test case with random initial positions. Note that my test will do that with the default case, too, so maybe someone should do that for the original test of the default implementation, too, to stay in sync.

**Comment by mkarg** [comment]
@bplb @AlanBateman As discussed before, I have changed the test in this PR to cover the following cases:
* Testing random buffer sizes
* Testing random initial source position
* Testing random initial target position

I deliberately did not test the following:
* 2GB+: IMHO makes not much sense to add such a heavy weight test as the expected benefit will not really outweigh the test complexity needed, will slow down test performance massively, and was not tested in the original implementation / in https://github.com/openjdk/jdk/blob/master/test/jdk/java/io/FileInputStream/TransferTo.java, too. There is no rule that we have to provide 100% test coverage in OpenJDK. Actually adding a test must be well justified and is not the only measure we have in the box to prevent the occurance of bugs in production: I suggest, instead of a 2GB+ test, we agree that the contributed implementation *looks* correct to each of us (= applying six-eyes principle). For such a simple code line, this should be enough QA, as in case it unexpectedly would break, it is very likely to get reported by beta testers long before GA (possibly by myself as many of my projects move around 4GB+ sized files in reality), which should be early enough to catch and fix.
* Same for testing position-after-transfer. This is not covered by the original test, and due to our six eyes it is highly unlikely that it is broken or will break in future. So I'd say KISS here.

Wherever it made sense (without sacrifying my intended modularity needed for other channel types tested later) I tried to follow the solutions found in https://github.com/openjdk/jdk/blob/master/test/jdk/java/io/FileInputStream/TransferTo.java. Thanks Bran, for this blue print. In fact I wonder if the introduction of 10 loops over random sizes actually will improve test qualify. My assumption is that it rather reduces test stability and test repeatability: In case there actually is a bug that happens only with a specific combination of buffer size and initial positions, ten iterations will be much too few to stably reproduce the bug at each execution of the test. More often that not, that ten loops will all pass without running into the problem given the huge max size. This will driver testers nuts! The improved entropy by introducing the JDK randomizer does not help much to improve this. Having said that, I copied the solution from Brian only because he came up with it, and I can live with the statistical unpredictability it induces, but just wanted to clearly say that I (spoiler: working for a company providing QA tools like FMEA and SPC software) would say introducing random here induces more harm due to non-repeatability than it actualy improves the detection rate of failures not detected by thorough six-eye review.

**Comment by AlanBateman** [comment]
> * 2GB+: IMHO makes not much sense to add such a heavy weight test as the expected benefit will not really outweigh the test complexity needed, 

I'll try to get time to look at what you have in the next few days but just to say that the suggestion for exercising files > 2GB is because it's beyond the limit of the underlying sendfile. You should be able to check this quickly by printing the iteration count in ChannelInputStream.transferTo, it is always 1?

**Review by mkarg — COMMENTED** [review]
> I'll try to get time to look at what you have in the next few days but just to say that the suggestion for exercising files > 2GB is because it's beyond the limit of the underlying sendfile. You should be able to check this quickly by printing the iteration count in ChannelInputStream.transferTo, it is always 1?

I understand your concerns and can dispel them. I added a counter that prints the number of iterations and yes, just as you expected, due to the `Long.MAX_VALUE` there is only a single iteration. But then I simply reduced `Long.MAX_VALUE` to `1024` and the counter reported thousands of iterations -- just as expected. Even better, the test confirmed that still the number of reported bytes is still correct just as the file contents are.

Hence in case sendfile *limits* the number of transferred bytes in production use, this quick check has proven that the code *is correct* and everything *actually* works fine.

**Comment by bplb** [comment]
As I understand it, if a `j.u.Random` instance is constructed from a given seed, then a series of invocations on that instance will always produce the same sequence of outputs. `RandomFactory` prints the seed in the test output and allows a seed to be provided to the test via a `-D` option. So if the test fails with a given combination of initial positions and buffer size, then it should be straightforward to reproduce the failure by providing the seed used when the test failed.

**Comment by mkarg** [comment]
> As I understand it, if a `j.u.Random` instance is constructed from a given seed, then a series of invocations on that instance will always produce the same sequence of outputs. `RandomFactory` prints the seed in the test output and allows a seed to be provided to the test via a `-D` option. So if the test fails with a given combination of initial positions and buffer size, then it should be straightforward to reproduce the failure by providing the seed used when the test failed.

Yes, given the fact it fails for ourselfs or at least for someone that knows that he has to tell us the printed seed.

**Comment by mkarg** [comment]
> The src/ implementation changes are fine, I think we agreed those previously.

Agreed.

> It would be cleaner if this test had been created as TestNG test. That would allow us to get rid of Thrower and the assertThrowsXXX infrastructure. It would allow us to use data providers too. Maybe it can be converted and cleaned up in a future PR.

I confirm that I will migrate to TestNG in a future PR.

> I think I mentioned this previously but the test can go into test/jdk/java/nio/channels/Channels so that it's in the same directory as the other tests for the Channels API/implementation.

Done in https://github.com/openjdk/jdk/pull/5209/commits/c886fbc26e18262539473bd66f091083fe9b7fbf.

> ifOutIsNullThenNpeIsThrown is a strange method name, can we rename it to something testNullPointerException or something more obvious?

Done in https://github.com/openjdk/jdk/pull/5209/commits/90a176f346d65782b703f194197c6858bce1077b.

> "contents" is also a bit strange for the method name. This where the real test is so let's rename that to something more obvious too.

Done in https://github.com/openjdk/jdk/pull/5209/commits/6b4c34c707778c68d3b06cc401c8c175918a60ea.

>There's a typo in one of the comments here it says "starting point", should be "starting position".

Done in https://github.com/openjdk/jdk/pull/5209/commits/cb2a863665f3fb329bd9c9237fc634a046d2edf1.

> You've got reasonable coverage for random initial positions in both source and target files, good! One other case that could be tested is where the initial position is beyond EOF. If the initial position of the source file is beyond EOF then 0 bytes will be transferred. If the initial position of the target file is beyond EOF then it will be extended.

Done in https://github.com/openjdk/jdk/pull/5209/commits/852032d9bbe102ebec1b03de21492221003180cd.

>At some point we should look at adding a test for files > 2Gb as otherwise we aren't testing the loop (as we discussed).

Agreed that *at some point* (i. e. in *some* future PR) someone could provide such a test. As explained in https://github.com/openjdk/jdk/pull/5209#issuecomment-907782217 there is no 100% path coverage rule in OpenJDK, so while this *might* be a good idea, it seems it is definitively not a *mandatory* constraint to test that loop at all.

> A few minor nits: The copyright date says "2014", I assume copied from another test.

It actually reads "2014, 2021" which means that the original code was first published in 2014 and was last modified in 2021. Oracle once asked me to always do it *exactly that way* when touching any Oracle source code (which I did here, as the test originated from existing Oracle source code). If OpenJDK is an exceptional case here, then please confirm that I am *officially allowed* to remove the year 2014 from that list. Until then I have to keep the line exactly the way it is to not get into any troubles with Oracle's legal department.

> There are several unused imports to remove.

Done in https://github.com/openjdk/jdk/pull/5209/commits/c3dbc33880ebea4b11f014c6c0046f3d5cc24e1e.

>The static modifiers on interfaces aren't needed.

Done in https://github.com/openjdk/jdk/pull/5209/commits/f3a6551d009d2a22d6d2aa448025330623c64065.

>There are few really long lines, one is 150+ that wraps and will be annoying for side-by-side diffs when changed in the future.

Done in https://github.com/openjdk/jdk/pull/5209/commits/412fab109c94de21a19cdc6a8aa3060e41deeb51. The longest code line now is 115 characters which should be fine for side-by-side views as my screen shows more than 250 characters per line.

I hope that with my recent changes this PR now is in a shape which allows it to get merged. As we agreed on multiple small PRs instead of a single big one, our target should be to contain only the *essential changes per each PR* but add *more PRs ontop* for non-essential improvements (which I will be happy to provide).

**Comment by AlanBateman** [comment]
> Agreed that _at some point_ (i. e. in _some_ future PR) someone could provide such a test. As explained in [#5209 (comment)](https://github.com/openjdk/jdk/pull/5209#issuecomment-907782217) there is no 100% path coverage rule in OpenJDK, so while this _might_ be a good idea, it seems it is definitively not a _mandatory_ constraint to test that loop at all.

The override of InputStream::transferTo is calling FileChannel::transferTo in a loop so I don't think it is unreasonable to ask if we have test that has more than one iteration. I'm okay if we defer this to a further step but I do think we should have test for this case.


> It actually reads "2014, 2021" which means that the original code was first published in 2014 and was last modified in 2021. Oracle once asked me to always do it _exactly that way_ when touching any Oracle source code (which I did here, as the test originated from existing Oracle source code). If OpenJDK is an exceptional case here, then please confirm that I am _officially allowed_ to remove the year 2014 from that list. Until then I have to keep the line exactly the way it is to not get into any troubles with Oracle's legal department.

This is a new test, the date should be 2021.

**Comment by mkarg** [comment]
> The override of InputStream::transferTo is calling FileChannel::transferTo in a loop so I don't think it is unreasonable to ask if we have test that has more than one iteration. I'm okay if we defer this to a further step but I do think we should have test for this case.

I will look into testing the 2 GB+ iterations after this PR is merged and I'm done with my TestNG PR.

> > It actually reads "2014, 2021" which means that the original code was first published in 2014 and was last modified in 2021. Oracle once asked me to always do it _exactly that way_ when touching any Oracle source code (which I did here, as the test originated from existing Oracle source code). If OpenJDK is an exceptional case here, then please confirm that I am _officially allowed_ to remove the year 2014 from that list. Until then I have to keep the line exactly the way it is to not get into any troubles with Oracle's legal department.
> This is a new test, the date should be 2021.

Done with https://github.com/openjdk/jdk/pull/5209/commits/1804aa00f34010a174301ac3d816de23806e179f.

So this PR is fine for merging now? :-)

**Comment by AlanBateman** [comment]
> Done with [1804aa0](https://github.com/openjdk/jdk/commit/1804aa00f34010a174301ac3d816de23806e179f).

Nearly there, it needs to be "2021," to avoid falling foul of scripts that check headers.

**Comment by mkarg** [comment]
> > Done with [1804aa0](https://github.com/openjdk/jdk/commit/1804aa00f34010a174301ac3d816de23806e179f).

> Nearly there, it needs to be "2021," to avoid falling foul of scripts that check headers.

Done with https://github.com/openjdk/jdk/pull/5209/commits/1f5a4567e3eaf55bf82f97164fb97c2abba481c8.

**Review by AlanBateman — APPROVED** [review]
I've tested the patch against the current main line (the PR is against a repo that is 4 or 5 months behind the main line). All okay. The expectation is that the test will be improved and iterated on in subsequent PRs.

**Pull request closed without merging**